Project using architectury loom fails to run a forge game instance in 1.15.2

When running `Minecraft Client (:forge)`, the game fails to finish loading with the following error: 
> Forge (forge) has failed to load correctly

A full log is available here: https://gist.github.com/Trikzon/df7831a9ac55de6c62b140b8deb97a5b

When the mod is built, the built jar works in a non-dev environment.
This is the workspace I am using and getting the error with: https://github.com/trikzon/armor-visibility/tree/1.15.2

> ```
> Caused by: java.lang.RuntimeException: Missing forge version, cannot continue
> 	at net.minecraftforge.versions.forge.ForgeVersion.<clinit>(ForgeVersion.java:45) ~[?:?]
> ```

This might be related to architectury/architectury-loom#16 🤔

--- a/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
@@ -34,6 +34,8 @@ import org.gradle.api.initialization.Settings;
 import org.gradle.api.invocation.Gradle;
 import org.gradle.api.plugins.PluginAware;
 
+import net.fabricmc.loom.util.Constants;
+
 public class LoomRepositoryPlugin implements Plugin<PluginAware> {
 	@Override
 	public void apply(PluginAware target) {
@@ -66,6 +68,15 @@ public class LoomRepositoryPlugin implements Plugin<PluginAware> {
 		repositories.maven(repo -> {
 			repo.setName("UserLocalRemappedMods");
 			repo.setUrl(cache.getRemappedModCache());
+		});
+		repositories.maven(repo -> {
+			repo.setName("Architectury");
+			repo.setUrl("https://maven.architectury.dev/");
+			repo.mavenContent(content -> {
+				content.releasesOnly();
+				String[] dep = Constants.Dependencies.FORGE_RUNTIME.split(":");
+				content.includeVersion(dep[0], dep[1], Constants.Dependencies.Versions.FORGE_RUNTIME);
+			});
 		});
 		repositories.maven(repo -> {
 			repo.setName("Fabric");

--- a/src/main/java/net/fabricmc/loom/build/mixin/AnnotationProcessorInvoker.java
+++ b/src/main/java/net/fabricmc/loom/build/mixin/AnnotationProcessorInvoker.java
@@ -99,6 +99,10 @@ public abstract class AnnotationProcessorInvoker<T extends Task> {
 								configs.getByName(Constants.Configurations.MAPPINGS_FINAL)
 				);
 
+				if (extension.isForge()) {
+					processorConfig.extendsFrom(configs.getByName(Constants.Configurations.FORGE_NAMED));
+				}
+
 				// Add Mixin and mixin extensions (fabric-mixin-compile-extensions pulls mixin itself too)
 				project.getDependencies().add(processorConfig.getName(),
 								Constants.Dependencies.MIXIN_COMPILE_EXTENSIONS + Constants.Dependencies.Versions.MIXIN_COMPILE_EXTENSIONS);

--- a/src/main/java/net/fabricmc/loom/configuration/CompileConfiguration.java
+++ b/src/main/java/net/fabricmc/loom/configuration/CompileConfiguration.java
@@ -132,10 +132,6 @@ public final class CompileConfiguration {
 		extendsFrom(Constants.Configurations.LOADER_DEPENDENCIES, Constants.Configurations.MINECRAFT_DEPENDENCIES, project);
 		extendsFrom(Constants.Configurations.MINECRAFT_NAMED, Constants.Configurations.LOADER_DEPENDENCIES, project);
 
-		if (project.getExtensions().getByType(LoomGradleExtension.class).isForge()) {
-			extendsFrom(Constants.Configurations.FORGE_NAMED, Constants.Configurations.LOADER_DEPENDENCIES, project);
-		}
-
 		extendsFrom(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME, Constants.Configurations.MAPPINGS_FINAL, project);
 
 		extendsFrom(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME, Constants.Configurations.LOOM_DEVELOPMENT_DEPENDENCIES, project);

--- a/src/main/java/net/fabricmc/loom/configuration/CompileConfiguration.java
+++ b/src/main/java/net/fabricmc/loom/configuration/CompileConfiguration.java
@@ -84,9 +84,15 @@ public final class CompileConfiguration {
 			data.createLazyConfiguration(Constants.Configurations.FORGE_INSTALLER).configure(configuration -> configuration.setTransitive(false));
 			data.createLazyConfiguration(Constants.Configurations.FORGE_UNIVERSAL).configure(configuration -> configuration.setTransitive(false));
 			data.createLazyConfiguration(Constants.Configurations.FORGE_DEPENDENCIES).configure(configuration -> configuration.setTransitive(false));
+			data.createLazyConfiguration(Constants.Configurations.FORGE_NAMED).configure(configuration -> configuration.setTransitive(false));
 			data.createLazyConfiguration(Constants.Configurations.MCP_CONFIG).configure(configuration -> configuration.setTransitive(false));
 
 			extendsFrom(Constants.Configurations.MINECRAFT_DEPENDENCIES, Constants.Configurations.FORGE_DEPENDENCIES, project);
+
+			extendsFrom(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME, Constants.Configurations.FORGE_NAMED, project);
+			extendsFrom(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME, Constants.Configurations.FORGE_NAMED, project);
+			extendsFrom(JavaPlugin.TEST_COMPILE_CLASSPATH_CONFIGURATION_NAME, Constants.Configurations.FORGE_NAMED, project);
+			extendsFrom(JavaPlugin.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME, Constants.Configurations.FORGE_NAMED, project);
 		}
 
 		if (project.getExtensions().getByType(LoomGradleExtension.class).supportsInclude()) {
@@ -125,6 +131,10 @@ public final class CompileConfiguration {
 
 		extendsFrom(Constants.Configurations.LOADER_DEPENDENCIES, Constants.Configurations.MINECRAFT_DEPENDENCIES, project);
 		extendsFrom(Constants.Configurations.MINECRAFT_NAMED, Constants.Configurations.LOADER_DEPENDENCIES, project);
+
+		if (project.getExtensions().getByType(LoomGradleExtension.class).isForge()) {
+			extendsFrom(Constants.Configurations.FORGE_NAMED, Constants.Configurations.LOADER_DEPENDENCIES, project);
+		}
 
 		extendsFrom(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME, Constants.Configurations.MAPPINGS_FINAL, project);
 

--- a/src/main/java/net/fabricmc/loom/configuration/mods/ModProcessor.java
+++ b/src/main/java/net/fabricmc/loom/configuration/mods/ModProcessor.java
@@ -163,6 +163,11 @@ public class ModProcessor {
 				.build();
 
 		remapper.readClassPathAsync(mc);
+
+		if (extension.isForge()) {
+			remapper.readClassPathAsync(mappedProvider.getForgeSrgJar().toPath());
+		}
+
 		remapper.readClassPathAsync(mcDeps);
 
 		final Map<ModDependencyInfo, InputTag> tagMap = new HashMap<>();

--- a/src/main/java/net/fabricmc/loom/configuration/providers/LaunchProvider.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/LaunchProvider.java
@@ -177,6 +177,10 @@ public class LaunchProvider extends DependencyProvider {
 
 		remapClasspath.add(getExtension().getMinecraftMappedProvider().getIntermediaryJar());
 
+		if (getExtension().isForge()) {
+			remapClasspath.add(getExtension().getMinecraftMappedProvider().getForgeIntermediaryJar());
+		}
+
 		String str = remapClasspath.stream()
 				.map(File::getAbsolutePath)
 				.collect(Collectors.joining(File.pathSeparator));

--- a/src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
@@ -58,6 +58,7 @@ import java.util.stream.Stream;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Predicates;
 import com.google.common.base.Stopwatch;
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.hash.Hashing;
 import com.google.common.io.ByteSource;
@@ -96,7 +97,7 @@ import net.fabricmc.mapping.tree.TinyTree;
 
 public class MinecraftPatchedProvider extends DependencyProvider {
 	private static final String LOOM_PATCH_VERSION_KEY = "Loom-Patch-Version";
-	private static final String CURRENT_LOOM_PATCH_VERSION = "3";
+	private static final String CURRENT_LOOM_PATCH_VERSION = "4";
 	private static final String NAME_MAPPING_SERVICE_PATH = "/inject/META-INF/services/cpw.mods.modlauncher.api.INameMappingService";
 
 	// Step 1: Remap Minecraft to SRG (global)
@@ -111,6 +112,8 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 	private File minecraftMergedPatchedSrgAtJar;
 	// Step 5: Remap Patched AT to Official (global or project)
 	private File minecraftMergedPatchedJar;
+	// Step 6: Remap Forge to Official (global)
+	private File forgeMergedJar;
 
 	private File projectAtHash;
 	private Set<File> projectAts = new HashSet<>();
@@ -177,6 +180,7 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 		minecraftClientPatchedSrgJar = new File(globalDir, "client-srg-patched.jar");
 		minecraftServerPatchedSrgJar = new File(globalDir, "server-srg-patched.jar");
 		minecraftMergedPatchedSrgJar = new File(globalDir, "merged-srg-patched.jar");
+		forgeMergedJar = new File(globalDir, "forge-official.jar");
 		minecraftMergedPatchedSrgAtJar = new File(projectDir, "merged-srg-at-patched.jar");
 		minecraftMergedPatchedJar = new File(projectDir, "merged-patched.jar");
 
@@ -214,6 +218,7 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 				minecraftClientPatchedSrgJar,
 				minecraftServerPatchedSrgJar,
 				minecraftMergedPatchedSrgJar,
+				forgeMergedJar,
 		};
 	}
 
@@ -251,7 +256,6 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 		if (!minecraftClientPatchedSrgJar.exists() || !minecraftServerPatchedSrgJar.exists()) {
 			this.dirty = true;
 			patchJars(getProject().getLogger());
-			injectForgeClasses(getProject().getLogger());
 		}
 	}
 
@@ -265,12 +269,50 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 			accessTransformForge(getProject().getLogger());
 		}
 
+		Path input = minecraftMergedPatchedSrgAtJar.toPath();
+		TinyRemapper remapper = null;
+
 		if (dirty) {
-			remapPatchedJar(getProject().getLogger());
+			getProject().getLogger().lifecycle(":remapping minecraft (TinyRemapper, srg -> official)");
+			remapper = buildRemapper(input);
+
+			remapPatchedJar(input, remapper, getProject().getLogger());
+		}
+
+		if (!forgeMergedJar.exists()) {
+			getProject().getLogger().lifecycle(":remapping forge (TinyRemapper, srg -> official)");
+
+			if (remapper != null) {
+				remapper = buildRemapper(input);
+			}
+
+			remapForgeJar(remapper, getProject().getLogger());
+		}
+
+		if (remapper != null) {
+			remapper.finish();
 		}
 
 		this.filesDirty = dirty;
 		this.dirty = false;
+	}
+
+	private TinyRemapper buildRemapper(Path input) throws IOException {
+		Path[] libraries = MinecraftMappedProvider.getRemapClasspath(getProject());
+		TinyTree mappingsWithSrg = getExtension().getMappingsProvider().getMappingsWithSrg();
+		TinyRemapper remapper = TinyRemapper.newRemapper()
+				.logger(getProject().getLogger()::lifecycle)
+				.logUnknownInvokeDynamic(false)
+				.withMappings(TinyRemapperMappingsHelper.create(mappingsWithSrg, "srg", "official", true))
+				.withMappings(InnerClassRemapper.of(input, mappingsWithSrg, "srg", "official"))
+				.renameInvalidLocals(true)
+				.rebuildSourceFilenames(true)
+				.fixPackageAccess(true)
+				.build();
+
+		remapper.readClassPath(libraries);
+		remapper.prepareClasses();
+		return remapper;
 	}
 
 	private void writeAtHash() throws IOException {
@@ -349,12 +391,12 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 		getProject().getLogger().info(":fixing parameter annotations for " + jarFile.getAbsolutePath() + " in " + stopwatch);
 	}
 
-	private void injectForgeClasses(Logger logger) throws IOException {
-		logger.lifecycle(":injecting forge classes into minecraft");
-		ThreadingUtils.run(Environment.values(), environment -> {
-			copyAll(getExtension().getForgeUniversalProvider().getForge(), environment.patchedSrgJar.apply(this));
-			copyUserdevFiles(getExtension().getForgeUserdevProvider().getUserdevJar(), environment.patchedSrgJar.apply(this));
-		});
+	private File getForgeJar() {
+		return getExtension().getForgeUniversalProvider().getForge();
+	}
+
+	private File getForgeUserdevJar() {
+		return getExtension().getForgeUserdevProvider().getUserdevJar();
 	}
 
 	private boolean isPatchedJarUpToDate(File jar) {
@@ -380,6 +422,7 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 	}
 
 	private void accessTransformForge(Logger logger) throws Exception {
+		List<File> toDelete = new ArrayList<>();
 		var atDependency = Constants.Dependencies.ACCESS_TRANSFORMERS + Constants.Dependencies.Versions.ACCESS_TRANSFORMERS;
 		var classpath = DependencyDownloader.download(getProject(), atDependency);
 
@@ -388,17 +431,27 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 		File input = minecraftMergedPatchedSrgJar;
 		File target = minecraftMergedPatchedSrgAtJar;
 		Files.deleteIfExists(target.toPath());
-		File at = File.createTempFile("at-conf", ".cfg");
-		at.deleteOnExit();
-		JarUtil.extractFile(input, "META-INF/accesstransformer.cfg", at);
 
 		List<String> args = new ArrayList<>();
 		args.add("--inJar");
 		args.add(input.getAbsolutePath());
 		args.add("--outJar");
 		args.add(target.getAbsolutePath());
-		args.add("--atFile");
-		args.add(at.getAbsolutePath());
+
+		for (File jar : ImmutableList.of(getForgeJar(), getForgeUserdevJar(), minecraftMergedPatchedSrgJar)) {
+			try (FileSystemUtil.FileSystemDelegate fs = FileSystemUtil.getJarFileSystem(jar, false)) {
+				Path atPath = fs.get().getPath("META-INF/accesstransformer.cfg");
+
+				if (Files.exists(atPath)) {
+					File tmpFile = File.createTempFile("at-conf", ".cfg");
+					tmpFile.delete();
+					toDelete.add(tmpFile);
+					Files.copy(atPath, tmpFile.toPath());
+					args.add("--atFile");
+					args.add(tmpFile.getAbsolutePath());
+				}
+			}
+		}
 
 		if (usesProjectCache()) {
 			for (File projectAt : projectAts) {
@@ -420,6 +473,10 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 				spec.setErrorOutput(NullOutputStream.NULL_OUTPUT_STREAM);
 			}
 		}).rethrowFailure().assertNormalExitValue();
+
+		for (File file : toDelete) {
+			file.delete();
+		}
 	}
 
 	public enum Environment {
@@ -444,34 +501,34 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 		}
 	}
 
-	private void remapPatchedJar(Logger logger) throws Exception {
-		Path[] libraries = MinecraftMappedProvider.getRemapClasspath(getProject());
-		logger.lifecycle(":remapping minecraft (TinyRemapper, srg -> official)");
-		TinyTree mappingsWithSrg = getExtension().getMappingsProvider().getMappingsWithSrg();
-
-		Path input = minecraftMergedPatchedSrgAtJar.toPath();
+	private void remapPatchedJar(Path input, TinyRemapper remapper, Logger logger) throws Exception {
 		Path output = minecraftMergedPatchedJar.toPath();
-
 		Files.deleteIfExists(output);
-
-		TinyRemapper remapper = TinyRemapper.newRemapper()
-				.logger(getProject().getLogger()::lifecycle)
-				.logUnknownInvokeDynamic(false)
-					.withMappings(TinyRemapperMappingsHelper.create(mappingsWithSrg, "srg", "official", true))
-				.withMappings(InnerClassRemapper.of(input, mappingsWithSrg, "srg", "official"))
-				.renameInvalidLocals(true)
-				.rebuildSourceFilenames(true)
-				.fixPackageAccess(true)
-				.build();
 
 		try (OutputConsumerPath outputConsumer = new OutputConsumerPath.Builder(output).build()) {
 			outputConsumer.addNonClassFiles(input);
 
-			remapper.readClassPath(libraries);
 			remapper.readInputs(input);
 			remapper.apply(outputConsumer);
 		} finally {
-			remapper.finish();
+			remapper.removeInput();
+		}
+
+		applyLoomPatchVersion(output);
+	}
+
+	private void remapForgeJar(TinyRemapper remapper, Logger logger) throws Exception {
+		Path output = forgeMergedJar.toPath();
+		Files.deleteIfExists(output);
+
+		try (OutputConsumerPath outputConsumer = new OutputConsumerPath.Builder(output).build()) {
+			Path forgeJar = getForgeJar().toPath();
+			Path forgeUserdevJar = getForgeUserdevJar().toPath();
+			outputConsumer.addNonClassFiles(forgeJar);
+			outputConsumer.addNonClassFiles(forgeUserdevJar);
+
+			remapper.readInputs(forgeJar, forgeUserdevJar);
+			remapper.apply(outputConsumer);
 		}
 	}
 
@@ -574,7 +631,7 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 	private void copyNonClassFiles(File source, File target) throws IOException {
 		Predicate<Path> filter = file -> {
 			String s = file.toString();
-			return !s.endsWith(".class") && !s.equalsIgnoreCase("/META-INF/MANIFEST.MF");
+			return !s.endsWith(".class");
 		};
 
 		walkFileSystems(source, target, filter, this::copyReplacing);
@@ -606,16 +663,20 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 
 			Files.copy(sourcePath, targetPath);
 		});
+	}
 
+	public void applyLoomPatchVersion(Path target) throws IOException {
 		try (FileSystemUtil.FileSystemDelegate delegate = FileSystemUtil.getJarFileSystem(target, false)) {
 			Path manifestPath = delegate.get().getPath("META-INF/MANIFEST.MF");
 
 			Preconditions.checkArgument(Files.exists(manifestPath), "META-INF/MANIFEST.MF does not exist in patched srg jar!");
 			Manifest manifest = new Manifest();
 
-			try (InputStream stream = Files.newInputStream(manifestPath)) {
-				manifest.read(stream);
-				manifest.getMainAttributes().putValue(LOOM_PATCH_VERSION_KEY, CURRENT_LOOM_PATCH_VERSION);
+			if (Files.exists(manifestPath)) {
+				try (InputStream stream = Files.newInputStream(manifestPath)) {
+					manifest.read(stream);
+					manifest.getMainAttributes().putValue(LOOM_PATCH_VERSION_KEY, CURRENT_LOOM_PATCH_VERSION);
+				}
 			}
 
 			try (OutputStream stream = Files.newOutputStream(manifestPath)) {
@@ -626,6 +687,10 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 
 	public File getMergedJar() {
 		return minecraftMergedPatchedJar;
+	}
+
+	public File getForgeMergedJar() {
+		return forgeMergedJar;
 	}
 
 	public boolean usesProjectCache() {

--- a/src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
@@ -64,6 +64,7 @@ import com.google.common.hash.Hashing;
 import com.google.common.io.ByteSource;
 import com.google.gson.JsonParser;
 import de.oceanlabs.mcp.mcinjector.adaptors.ParameterAnnotationFixer;
+import dev.architectury.tinyremapper.InputTag;
 import dev.architectury.tinyremapper.OutputConsumerPath;
 import dev.architectury.tinyremapper.TinyRemapper;
 import net.minecraftforge.binarypatcher.ConsoleTool;
@@ -87,7 +88,6 @@ import net.fabricmc.loom.configuration.providers.minecraft.MinecraftMappedProvid
 import net.fabricmc.loom.util.Constants;
 import net.fabricmc.loom.util.DependencyDownloader;
 import net.fabricmc.loom.util.FileSystemUtil;
-import net.fabricmc.loom.util.JarUtil;
 import net.fabricmc.loom.util.ThreadingUtils;
 import net.fabricmc.loom.util.TinyRemapperMappingsHelper;
 import net.fabricmc.loom.util.function.FsPathConsumer;
@@ -110,9 +110,8 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 	private File minecraftMergedPatchedSrgJar;
 	// Step 4: Access Transform (global or project)
 	private File minecraftMergedPatchedSrgAtJar;
-	// Step 5: Remap Patched AT to Official (global or project)
+	// Step 5: Remap Patched AT & Forge to Official (global or project)
 	private File minecraftMergedPatchedJar;
-	// Step 6: Remap Forge to Official (global)
 	private File forgeMergedJar;
 
 	private File projectAtHash;
@@ -269,28 +268,14 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 			accessTransformForge(getProject().getLogger());
 		}
 
-		Path input = minecraftMergedPatchedSrgAtJar.toPath();
-		TinyRemapper remapper = null;
+		if (!forgeMergedJar.exists()) {
+			this.dirty = true;
+		}
 
+		Path input = minecraftMergedPatchedSrgAtJar.toPath();
 		if (dirty) {
 			getProject().getLogger().lifecycle(":remapping minecraft (TinyRemapper, srg -> official)");
-			remapper = buildRemapper(input);
-
-			remapPatchedJar(input, remapper, getProject().getLogger());
-		}
-
-		if (!forgeMergedJar.exists()) {
-			getProject().getLogger().lifecycle(":remapping forge (TinyRemapper, srg -> official)");
-
-			if (remapper != null) {
-				remapper = buildRemapper(input);
-			}
-
-			remapForgeJar(remapper, getProject().getLogger());
-		}
-
-		if (remapper != null) {
-			remapper.finish();
+			remapPatchedJar(input, getProject().getLogger());
 		}
 
 		this.filesDirty = dirty;
@@ -501,35 +486,32 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 		}
 	}
 
-	private void remapPatchedJar(Path input, TinyRemapper remapper, Logger logger) throws Exception {
-		Path output = minecraftMergedPatchedJar.toPath();
-		Files.deleteIfExists(output);
+	private void remapPatchedJar(Path input, Logger logger) throws Exception {
+		Path mcOutput = minecraftMergedPatchedJar.toPath();
+		Path forgeOutput = forgeMergedJar.toPath();
+		Path forgeJar = getForgeJar().toPath();
+		Path forgeUserdevJar = getForgeUserdevJar().toPath();
+		Files.deleteIfExists(mcOutput);
+		Files.deleteIfExists(forgeOutput);
+		TinyRemapper remapper = buildRemapper(input);
 
-		try (OutputConsumerPath outputConsumer = new OutputConsumerPath.Builder(output).build()) {
+		try (OutputConsumerPath outputConsumer = new OutputConsumerPath.Builder(mcOutput).build();
+					OutputConsumerPath outputConsumerForge = new OutputConsumerPath.Builder(forgeOutput).build()) {
 			outputConsumer.addNonClassFiles(input);
 
-			remapper.readInputs(input);
-			remapper.apply(outputConsumer);
+			InputTag mcTag = remapper.createInputTag();
+			InputTag forgeTag = remapper.createInputTag();
+			remapper.readInputs(mcTag, input);
+			remapper.readInputs(forgeTag, forgeJar, forgeUserdevJar);
+			remapper.apply(outputConsumer, mcTag);
+			remapper.apply(outputConsumerForge, forgeTag);
 		} finally {
-			remapper.removeInput();
+			remapper.finish();
 		}
 
-		applyLoomPatchVersion(output);
-	}
-
-	private void remapForgeJar(TinyRemapper remapper, Logger logger) throws Exception {
-		Path output = forgeMergedJar.toPath();
-		Files.deleteIfExists(output);
-
-		try (OutputConsumerPath outputConsumer = new OutputConsumerPath.Builder(output).build()) {
-			Path forgeJar = getForgeJar().toPath();
-			Path forgeUserdevJar = getForgeUserdevJar().toPath();
-			outputConsumer.addNonClassFiles(forgeJar);
-			outputConsumer.addNonClassFiles(forgeUserdevJar);
-
-			remapper.readInputs(forgeJar, forgeUserdevJar);
-			remapper.apply(outputConsumer);
-		}
+		applyLoomPatchVersion(mcOutput);
+		copyNonClassFiles(forgeJar.toFile(), forgeMergedJar);
+		copyUserdevFiles(forgeUserdevJar.toFile(), forgeMergedJar);
 	}
 
 	private void patchJars(Logger logger) throws IOException {
@@ -652,7 +634,7 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 		// If there are multiple name mapping services with the same "understanding" pair
 		// (source -> target namespace pair), modlauncher throws a fit and will crash.
 		// To use our YarnNamingService instead of MCPNamingService, we have to remove this file.
-		Predicate<Path> filter = file -> !file.toString().equals(NAME_MAPPING_SERVICE_PATH);
+		Predicate<Path> filter = file -> !file.toString().endsWith(".class") && !file.toString().equals(NAME_MAPPING_SERVICE_PATH);
 
 		walkFileSystems(source, target, filter, fs -> Collections.singleton(fs.getPath("inject")), (sourceFs, targetFs, sourcePath, targetPath) -> {
 			Path parent = targetPath.getParent();

--- a/src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java
@@ -48,6 +48,7 @@ import java.util.List;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Predicate;
@@ -274,7 +275,6 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 
 		Path input = minecraftMergedPatchedSrgAtJar.toPath();
 		if (dirty) {
-			getProject().getLogger().lifecycle(":remapping minecraft (TinyRemapper, srg -> official)");
 			remapPatchedJar(input, getProject().getLogger());
 		}
 
@@ -487,6 +487,7 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 	}
 
 	private void remapPatchedJar(Path input, Logger logger) throws Exception {
+		getProject().getLogger().lifecycle(":remapping minecraft (TinyRemapper, srg -> official)");
 		Path mcOutput = minecraftMergedPatchedJar.toPath();
 		Path forgeOutput = forgeMergedJar.toPath();
 		Path forgeJar = getForgeJar().toPath();
@@ -501,8 +502,10 @@ public class MinecraftPatchedProvider extends DependencyProvider {
 
 			InputTag mcTag = remapper.createInputTag();
 			InputTag forgeTag = remapper.createInputTag();
-			remapper.readInputs(mcTag, input);
-			remapper.readInputs(forgeTag, forgeJar, forgeUserdevJar);
+			List<CompletableFuture<?>> futures = new ArrayList<>();
+			futures.add(remapper.readInputsAsync(mcTag, input));
+			futures.add(remapper.readInputsAsync(forgeTag, forgeJar, forgeUserdevJar));
+			CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
 			remapper.apply(outputConsumer, mcTag);
 			remapper.apply(outputConsumerForge, forgeTag);
 		} finally {

--- a/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMappedProvider.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMappedProvider.java
@@ -69,9 +69,13 @@ public class MinecraftMappedProvider extends DependencyProvider {
 			.build();
 
 	private File inputJar;
+	private File inputForgeJar;
 	private File minecraftMappedJar;
 	private File minecraftIntermediaryJar;
 	private File minecraftSrgJar;
+	private File forgeMappedJar;
+	private File forgeIntermediaryJar;
+	private File forgeSrgJar;
 
 	private MinecraftProviderImpl minecraftProvider;
 
@@ -90,6 +94,7 @@ public class MinecraftMappedProvider extends DependencyProvider {
 		}
 
 		boolean isForgeAtDirty = getExtension().isForge() && getExtension().getMappingsProvider().patchedProvider.isAtDirty();
+		TinyRemapper[] remapperArray = new TinyRemapper[] {null};
 
 		if (!minecraftMappedJar.exists() || !getIntermediaryJar().exists() || (getExtension().isForge() && !getSrgJar().exists()) || isRefreshDeps() || isForgeAtDirty) {
 			if (minecraftMappedJar.exists()) {
@@ -107,7 +112,7 @@ public class MinecraftMappedProvider extends DependencyProvider {
 			}
 
 			try {
-				mapMinecraftJar();
+				mapMinecraftJar(remapperArray);
 			} catch (Throwable t) {
 				// Cleanup some some things that may be in a bad state now
 				DownloadUtil.delete(minecraftMappedJar);
@@ -115,12 +120,35 @@ public class MinecraftMappedProvider extends DependencyProvider {
 				getExtension().getMinecraftProvider().deleteFiles();
 
 				if (getExtension().isForge()) {
-					minecraftSrgJar.delete();
+					DownloadUtil.delete(minecraftSrgJar);
 				}
 
 				getExtension().getMappingsProvider().cleanFiles();
 				throw new RuntimeException("Failed to remap minecraft", t);
 			}
+		}
+
+		if (getExtension().isForge() && (!getForgeMappedJar().exists() || !getForgeIntermediaryJar().exists() || !getForgeSrgJar().exists() || isRefreshDeps() || isForgeAtDirty)) {
+			if (getForgeMappedJar().exists()) {
+				getForgeMappedJar().delete();
+			}
+
+			getForgeMappedJar().getParentFile().mkdirs();
+			getForgeIntermediaryJar().delete();
+			getForgeSrgJar().delete();
+
+			try {
+				mapForgeJar(remapperArray);
+			} catch (Throwable t) {
+				DownloadUtil.delete(forgeMappedJar);
+				DownloadUtil.delete(forgeSrgJar);
+				DownloadUtil.delete(forgeIntermediaryJar);
+				throw new RuntimeException("Failed to remap forge", t);
+			}
+		}
+
+		if (remapperArray[0] != null) {
+			remapperArray[0].finish();
 		}
 
 		if (!minecraftMappedJar.exists()) {
@@ -129,36 +157,31 @@ public class MinecraftMappedProvider extends DependencyProvider {
 
 		addDependencies(dependency, postPopulationScheduler);
 
-		getProject().afterEvaluate(project -> {
-			if (getExtension().isForge() && !OperatingSystem.isCIBuild()) {
-				try {
-					ForgeSourcesRemapper.addBaseForgeSources(project);
-				} catch (IOException e) {
-					e.printStackTrace();
+		if (getExtension().isForge()) {
+			getProject().getDependencies().add(Constants.Configurations.FORGE_NAMED,
+					getProject().getDependencies().module("net.minecraftforge-loom:forge:" + getJarVersionString("mapped")));
+
+			getProject().afterEvaluate(project -> {
+				if (!OperatingSystem.isCIBuild()) {
+					try {
+						ForgeSourcesRemapper.addBaseForgeSources(project);
+					} catch (IOException e) {
+						e.printStackTrace();
+					}
 				}
-			}
-		});
+			});
+		}
 	}
 
-	private void mapMinecraftJar() throws Exception {
-		String fromM = "official";
-
-		MappingsProviderImpl mappingsProvider = getExtension().getMappingsProvider();
-
-		Path input = inputJar.toPath();
-		Path outputMapped = minecraftMappedJar.toPath();
-		Path outputIntermediary = minecraftIntermediaryJar.toPath();
-		Path outputSrg = minecraftSrgJar == null ? null : minecraftSrgJar.toPath();
-
+	private TinyRemapper buildRemapper() throws IOException {
 		Path[] libraries = getRemapClasspath(getProject());
 		TinyRemapper remapper = getTinyRemapper();
 		remapper.readClassPath(libraries);
 		remapper.prepareClasses();
-
-		Path tmpAssets = Files.createTempFile("tmpAssets", null);
-		Files.deleteIfExists(tmpAssets);
-		tmpAssets.toFile().deleteOnExit();
-
+		return remapper;
+	}
+	
+	private byte[][] inputBytes(Path input, @Nullable Path assetsOut) throws IOException {
 		List<byte[]> inputByteList = new ArrayList<>();
 
 		try (FileSystemUtil.FileSystemDelegate inputFs = FileSystemUtil.getJarFileSystem(input, false)) {
@@ -181,38 +204,72 @@ public class MinecraftMappedProvider extends DependencyProvider {
 			taskCompleter.complete();
 		}
 
-		try (OutputConsumerPath tmpAssetsPath = new OutputConsumerPath.Builder(tmpAssets).assumeArchive(true).build()) {
-			if (getExtension().isForge()) {
-				tmpAssetsPath.addNonClassFiles(input, NonClassCopyMode.FIX_META_INF, null);
-			} else {
-				tmpAssetsPath.addNonClassFiles(input);
+		if (assetsOut != null) {
+			try (OutputConsumerPath tmpAssetsPath = new OutputConsumerPath.Builder(assetsOut).assumeArchive(true).build()) {
+				if (getExtension().isForge()) {
+					tmpAssetsPath.addNonClassFiles(input, NonClassCopyMode.FIX_META_INF, null);
+				} else {
+					tmpAssetsPath.addNonClassFiles(input);
+				}
 			}
 		}
 
-		byte[][] inputBytes = inputByteList.toArray(new byte[0][0]);
+		return inputByteList.toArray(new byte[0][0]);
+	}
 
+	private void mapMinecraftJar(TinyRemapper[] remapperArray) throws Exception {
+		Path input = inputJar.toPath();
+		Path outputMapped = minecraftMappedJar.toPath();
+		Path outputIntermediary = minecraftIntermediaryJar.toPath();
+		Path outputSrg = minecraftSrgJar == null ? null : minecraftSrgJar.toPath();
+		TinyRemapper remapper = remapperArray[0] = buildRemapper();
+
+		Path assets = Files.createTempFile("assets", null);
+		Files.deleteIfExists(assets);
+		assets.toFile().deleteOnExit();
+		byte[][] inputBytes = inputBytes(input, assets);
+		remap(remapper, "minecraft", inputBytes, assets, input, outputMapped,
+				outputIntermediary, outputSrg, "official");
+	}
+
+	private void mapForgeJar(TinyRemapper[] remapperArray) throws Exception {
+		Path input = inputJar.toPath();
+		Path inputForge = inputForgeJar.toPath();
+		Path outputMapped = forgeMappedJar.toPath();
+		Path outputIntermediary = forgeIntermediaryJar.toPath();
+		Path outputSrg = forgeSrgJar.toPath();
+		TinyRemapper remapper = remapperArray[0] != null ? remapperArray[0] : (remapperArray[0] = buildRemapper());
+
+		Path assets = Files.createTempFile("assets", null);
+		Files.deleteIfExists(assets);
+		assets.toFile().deleteOnExit();
+		byte[][] inputBytes = inputBytes(inputForge, assets);
+		remap(remapper, "forge", inputBytes, assets, input, outputMapped,
+				outputIntermediary, outputSrg, "official");
+	}
+	
+	public void remap(TinyRemapper remapper, String id, byte[][] inputBytes, Path assets, Path mcOfficialJar,
+			Path outputMapped, Path outputIntermediary, Path outputSrg, String fromM) throws IOException {
 		for (String toM : getExtension().isForge() ? Arrays.asList("intermediary", "srg", "named") : Arrays.asList("intermediary", "named")) {
 			Path output = "named".equals(toM) ? outputMapped : "srg".equals(toM) ? outputSrg : outputIntermediary;
 			Stopwatch stopwatch = Stopwatch.createStarted();
-			getProject().getLogger().lifecycle(":remapping minecraft (TinyRemapper, " + fromM + " -> " + toM + ")");
+			getProject().getLogger().lifecycle(":remapping " + id + " (TinyRemapper, " + fromM + " -> " + toM + ")");
 
 			remapper.readInputs(inputBytes);
-			remapper.replaceMappings(getMappings(input, fromM, toM));
-			OutputRemappingHandler.remap(remapper, tmpAssets, output);
+			remapper.replaceMappings(getMappings(mcOfficialJar, fromM, toM));
+			OutputRemappingHandler.remap(remapper, assets, output);
 
-			getProject().getLogger().lifecycle(":remapped minecraft (TinyRemapper, " + fromM + " -> " + toM + ") in " + stopwatch);
+			getProject().getLogger().lifecycle(":remapped " + id + " (TinyRemapper, " + fromM + " -> " + toM + ") in " + stopwatch);
 			remapper.removeInput();
 
 			if (getExtension().isForge() && !"srg".equals(toM)) {
-				getProject().getLogger().info(":running forge finalising tasks");
+				getProject().getLogger().info(":running " + id + " finalising tasks");
 
 				TinyTree yarnWithSrg = getExtension().getMappingsProvider().getMappingsWithSrg();
 				AtRemapper.remap(getProject().getLogger(), output, yarnWithSrg);
 				CoreModClassRemapper.remapJar(output, yarnWithSrg, getProject().getLogger());
 			}
 		}
-
-		remapper.finish();
 	}
 
 	public TinyRemapper getTinyRemapper() throws IOException {
@@ -266,6 +323,12 @@ public class MinecraftMappedProvider extends DependencyProvider {
 		minecraftSrgJar = !getExtension().isForge() ? null : new File(getExtension().getUserCache(), "minecraft-" + getJarVersionString("srg") + ".jar");
 		minecraftMappedJar = new File(getJarDirectory(getExtension().getUserCache(), "mapped"), "minecraft-" + getJarVersionString("mapped") + ".jar");
 		inputJar = getExtension().isForge() ? mappingsProvider.patchedProvider.getMergedJar() : minecraftProvider.getMergedJar();
+		if (getExtension().isForge()) {
+			inputForgeJar = mappingsProvider.patchedProvider.getForgeMergedJar();
+			forgeIntermediaryJar = new File(getExtension().getUserCache(), "forge-" + getJarVersionString("intermediary") + ".jar");
+			forgeSrgJar = new File(getExtension().getUserCache(), "forge-" + getJarVersionString("srg") + ".jar");
+			forgeMappedJar = new File(getJarDirectory(getExtension().getUserCache(), "mapped"), "forge-" + getJarVersionString("mapped") + ".jar");
+		}
 	}
 
 	protected File getJarDirectory(File parentDirectory, String type) {
@@ -286,6 +349,18 @@ public class MinecraftMappedProvider extends DependencyProvider {
 
 	public File getMappedJar() {
 		return minecraftMappedJar;
+	}
+
+	public File getForgeIntermediaryJar() {
+		return forgeIntermediaryJar;
+	}
+
+	public File getForgeSrgJar() {
+		return forgeSrgJar;
+	}
+
+	public File getForgeMappedJar() {
+		return forgeMappedJar;
 	}
 
 	public File getUnpickedJar() {

--- a/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMappedProvider.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMappedProvider.java
@@ -39,6 +39,7 @@ import java.util.function.Consumer;
 import com.google.common.base.Stopwatch;
 import com.google.common.collect.ImmutableMap;
 import dev.architectury.tinyremapper.IMappingProvider;
+import dev.architectury.tinyremapper.InputTag;
 import dev.architectury.tinyremapper.NonClassCopyMode;
 import dev.architectury.tinyremapper.OutputConsumerPath;
 import dev.architectury.tinyremapper.TinyRemapper;
@@ -94,9 +95,17 @@ public class MinecraftMappedProvider extends DependencyProvider {
 		}
 
 		boolean isForgeAtDirty = getExtension().isForge() && getExtension().getMappingsProvider().patchedProvider.isAtDirty();
-		TinyRemapper[] remapperArray = new TinyRemapper[] {null};
+		boolean needToRemap = false;
 
 		if (!minecraftMappedJar.exists() || !getIntermediaryJar().exists() || (getExtension().isForge() && !getSrgJar().exists()) || isRefreshDeps() || isForgeAtDirty) {
+			needToRemap = true;
+		}
+
+		if (getExtension().isForge() && (!getForgeMappedJar().exists() || !getForgeIntermediaryJar().exists() || !getForgeSrgJar().exists() || isRefreshDeps() || isForgeAtDirty)) {
+			needToRemap = true;
+		}
+
+		if (needToRemap) {
 			if (minecraftMappedJar.exists()) {
 				minecraftMappedJar.delete();
 			}
@@ -111,24 +120,6 @@ public class MinecraftMappedProvider extends DependencyProvider {
 				minecraftSrgJar.delete();
 			}
 
-			try {
-				mapMinecraftJar(remapperArray);
-			} catch (Throwable t) {
-				// Cleanup some some things that may be in a bad state now
-				DownloadUtil.delete(minecraftMappedJar);
-				DownloadUtil.delete(minecraftIntermediaryJar);
-				getExtension().getMinecraftProvider().deleteFiles();
-
-				if (getExtension().isForge()) {
-					DownloadUtil.delete(minecraftSrgJar);
-				}
-
-				getExtension().getMappingsProvider().cleanFiles();
-				throw new RuntimeException("Failed to remap minecraft", t);
-			}
-		}
-
-		if (getExtension().isForge() && (!getForgeMappedJar().exists() || !getForgeIntermediaryJar().exists() || !getForgeSrgJar().exists() || isRefreshDeps() || isForgeAtDirty)) {
 			if (getForgeMappedJar().exists()) {
 				getForgeMappedJar().delete();
 			}
@@ -138,17 +129,25 @@ public class MinecraftMappedProvider extends DependencyProvider {
 			getForgeSrgJar().delete();
 
 			try {
-				mapForgeJar(remapperArray);
+				TinyRemapper[] remapperArray = new TinyRemapper[] {null};
+				mapMinecraftJar(remapperArray);
+				remapperArray[0].finish();
 			} catch (Throwable t) {
-				DownloadUtil.delete(forgeMappedJar);
-				DownloadUtil.delete(forgeSrgJar);
-				DownloadUtil.delete(forgeIntermediaryJar);
-				throw new RuntimeException("Failed to remap forge", t);
-			}
-		}
+				// Cleanup some some things that may be in a bad state now
+				DownloadUtil.delete(minecraftMappedJar);
+				DownloadUtil.delete(minecraftIntermediaryJar);
+				getExtension().getMinecraftProvider().deleteFiles();
 
-		if (remapperArray[0] != null) {
-			remapperArray[0].finish();
+				if (getExtension().isForge()) {
+					DownloadUtil.delete(minecraftSrgJar);
+					DownloadUtil.delete(forgeMappedJar);
+					DownloadUtil.delete(forgeSrgJar);
+					DownloadUtil.delete(forgeIntermediaryJar);
+				}
+
+				getExtension().getMappingsProvider().cleanFiles();
+				throw new RuntimeException("Failed to remap minecraft", t);
+			}
 		}
 
 		if (!minecraftMappedJar.exists()) {
@@ -180,8 +179,8 @@ public class MinecraftMappedProvider extends DependencyProvider {
 		remapper.prepareClasses();
 		return remapper;
 	}
-	
-	private byte[][] inputBytes(Path input, @Nullable Path assetsOut) throws IOException {
+
+	private byte[][] inputBytes(Path input) throws IOException {
 		List<byte[]> inputByteList = new ArrayList<>();
 
 		try (FileSystemUtil.FileSystemDelegate inputFs = FileSystemUtil.getJarFileSystem(input, false)) {
@@ -204,6 +203,10 @@ public class MinecraftMappedProvider extends DependencyProvider {
 			taskCompleter.complete();
 		}
 
+		return inputByteList.toArray(new byte[0][0]);
+	}
+
+	private void assetsOut(Path input, @Nullable Path assetsOut) throws IOException {
 		if (assetsOut != null) {
 			try (OutputConsumerPath tmpAssetsPath = new OutputConsumerPath.Builder(assetsOut).assumeArchive(true).build()) {
 				if (getExtension().isForge()) {
@@ -213,57 +216,83 @@ public class MinecraftMappedProvider extends DependencyProvider {
 				}
 			}
 		}
-
-		return inputByteList.toArray(new byte[0][0]);
 	}
 
 	private void mapMinecraftJar(TinyRemapper[] remapperArray) throws Exception {
 		Path input = inputJar.toPath();
+		Path inputForge = inputForgeJar.toPath();
 		Path outputMapped = minecraftMappedJar.toPath();
 		Path outputIntermediary = minecraftIntermediaryJar.toPath();
 		Path outputSrg = minecraftSrgJar == null ? null : minecraftSrgJar.toPath();
+
+		Path forgeOutputMapped = forgeMappedJar == null ? null : forgeMappedJar.toPath();
+		Path forgeOutputIntermediary = forgeIntermediaryJar == null ? null : forgeIntermediaryJar.toPath();
+		Path forgeOutputSrg = forgeSrgJar == null ? null : forgeSrgJar.toPath();
+
+		Path vanillaAssets = Files.createTempFile("assets", null);
+		Files.deleteIfExists(vanillaAssets);
+		vanillaAssets.toFile().deleteOnExit();
+		Path forgeAssets = Files.createTempFile("assets", null);
+		Files.deleteIfExists(forgeAssets);
+		forgeAssets.toFile().deleteOnExit();
+
+		Info vanilla = new Info(vanillaAssets, input, outputMapped, outputIntermediary, outputSrg);
+		Info forge = getExtension().isForge() ? new Info(forgeAssets, inputForge, forgeOutputMapped, forgeOutputIntermediary, forgeOutputSrg) : null;
+
 		TinyRemapper remapper = remapperArray[0] = buildRemapper();
 
-		Path assets = Files.createTempFile("assets", null);
-		Files.deleteIfExists(assets);
-		assets.toFile().deleteOnExit();
-		byte[][] inputBytes = inputBytes(input, assets);
-		remap(remapper, "minecraft", inputBytes, assets, input, outputMapped,
-				outputIntermediary, outputSrg, "official");
+		assetsOut(input, vanillaAssets);
+
+		if ( getExtension().isForge()) {
+			assetsOut(inputForge, forgeAssets);
+		}
+
+		remap(remapper, vanilla, forge, "official");
 	}
 
-	private void mapForgeJar(TinyRemapper[] remapperArray) throws Exception {
-		Path input = inputJar.toPath();
-		Path inputForge = inputForgeJar.toPath();
-		Path outputMapped = forgeMappedJar.toPath();
-		Path outputIntermediary = forgeIntermediaryJar.toPath();
-		Path outputSrg = forgeSrgJar.toPath();
-		TinyRemapper remapper = remapperArray[0] != null ? remapperArray[0] : (remapperArray[0] = buildRemapper());
+	public static class Info {
+		Path assets;
+		Path input;
+		Path outputMapped;
+		Path outputIntermediary;
+		Path outputSrg;
 
-		Path assets = Files.createTempFile("assets", null);
-		Files.deleteIfExists(assets);
-		assets.toFile().deleteOnExit();
-		byte[][] inputBytes = inputBytes(inputForge, assets);
-		remap(remapper, "forge", inputBytes, assets, input, outputMapped,
-				outputIntermediary, outputSrg, "official");
+		public Info(Path assets, Path input, Path outputMapped, Path outputIntermediary, Path outputSrg) {
+			this.assets = assets;
+			this.input = input;
+			this.outputMapped = outputMapped;
+			this.outputIntermediary = outputIntermediary;
+			this.outputSrg = outputSrg;
+		}
 	}
-	
-	public void remap(TinyRemapper remapper, String id, byte[][] inputBytes, Path assets, Path mcOfficialJar,
-			Path outputMapped, Path outputIntermediary, Path outputSrg, String fromM) throws IOException {
+
+	public void remap(TinyRemapper remapper, Info vanilla, @Nullable Info forge, String fromM) throws IOException {
 		for (String toM : getExtension().isForge() ? Arrays.asList("intermediary", "srg", "named") : Arrays.asList("intermediary", "named")) {
-			Path output = "named".equals(toM) ? outputMapped : "srg".equals(toM) ? outputSrg : outputIntermediary;
+			Path output = "named".equals(toM) ? vanilla.outputMapped : "srg".equals(toM) ? vanilla.outputSrg : vanilla.outputIntermediary;
+			Path outputForge = forge == null ? null : "named".equals(toM) ? forge.outputMapped : "srg".equals(toM) ? forge.outputSrg : forge.outputIntermediary;
+			InputTag vanillaTag = remapper.createInputTag();
+			InputTag forgeTag = remapper.createInputTag();
 			Stopwatch stopwatch = Stopwatch.createStarted();
-			getProject().getLogger().lifecycle(":remapping " + id + " (TinyRemapper, " + fromM + " -> " + toM + ")");
+			getProject().getLogger().lifecycle(":remapping minecraft (TinyRemapper, " + fromM + " -> " + toM + ")");
 
-			remapper.readInputs(inputBytes);
-			remapper.replaceMappings(getMappings(mcOfficialJar, fromM, toM));
-			OutputRemappingHandler.remap(remapper, assets, output);
+			remapper.readInputs(vanillaTag, vanilla.input);
 
-			getProject().getLogger().lifecycle(":remapped " + id + " (TinyRemapper, " + fromM + " -> " + toM + ") in " + stopwatch);
+			if (forge != null) {
+				remapper.readInputs(forgeTag, forge.input);
+			}
+
+			remapper.replaceMappings(getMappings(vanilla.input, fromM, toM));
+			OutputRemappingHandler.remap(remapper, vanilla.assets, output, null, vanillaTag);
+
+			if (forge != null) {
+				OutputRemappingHandler.remap(remapper, forge.assets, outputForge, null, forgeTag);
+			}
+
+			getProject().getLogger().lifecycle(":remapped minecraft (TinyRemapper, " + fromM + " -> " + toM + ") in " + stopwatch);
 			remapper.removeInput();
 
 			if (getExtension().isForge() && !"srg".equals(toM)) {
-				getProject().getLogger().info(":running " + id + " finalising tasks");
+				getProject().getLogger().info(":running minecraft finalising tasks");
 
 				TinyTree yarnWithSrg = getExtension().getMappingsProvider().getMappingsWithSrg();
 				AtRemapper.remap(getProject().getLogger(), output, yarnWithSrg);

--- a/src/main/java/net/fabricmc/loom/configuration/sources/ForgeSourcesRemapper.java
+++ b/src/main/java/net/fabricmc/loom/configuration/sources/ForgeSourcesRemapper.java
@@ -59,7 +59,7 @@ import net.fabricmc.lorenztiny.TinyMappingsReader;
 
 public class ForgeSourcesRemapper {
 	public static void addBaseForgeSources(Project project) throws IOException {
-		Path sourcesJar = GenerateSourcesTask.getMappedJarFileWithSuffix(project, "-sources.jar").toPath();
+		Path sourcesJar = GenerateSourcesTask.getMappedJarFileWithSuffix(project, "-sources.jar", true).toPath();
 
 		if (!Files.exists(sourcesJar)) {
 			addForgeSources(project, sourcesJar);
@@ -185,6 +185,7 @@ public class ForgeSourcesRemapper {
 
 		// Distinct and add the srg jar at the top, so it gets prioritized
 		mercury.getClassPath().add(0, extension.getMinecraftMappedProvider().getSrgJar().toPath());
+		mercury.getClassPath().add(0, extension.getMinecraftMappedProvider().getForgeSrgJar().toPath());
 		List<Path> newClassPath = mercury.getClassPath().stream()
 				.distinct()
 				.filter(Files::isRegularFile)

--- a/src/main/java/net/fabricmc/loom/task/GenerateSourcesTask.java
+++ b/src/main/java/net/fabricmc/loom/task/GenerateSourcesTask.java
@@ -110,9 +110,13 @@ public class GenerateSourcesTask extends AbstractLoomTask {
 	}
 
 	public static File getMappedJarFileWithSuffix(Project project, String suffix) {
+		return getMappedJarFileWithSuffix(project, suffix, false);
+	}
+
+	public static File getMappedJarFileWithSuffix(Project project, String suffix, boolean forgeJar) {
 		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
 		MappingsProviderImpl mappingsProvider = extension.getMappingsProvider();
-		File mappedJar = mappingsProvider.mappedProvider.getMappedJar();
+		File mappedJar = forgeJar ? mappingsProvider.mappedProvider.getForgeMappedJar() : mappingsProvider.mappedProvider.getMappedJar();
 		String path = mappedJar.getAbsolutePath();
 
 		if (!path.toLowerCase(Locale.ROOT).endsWith(".jar")) {

--- a/src/main/java/net/fabricmc/loom/task/MigrateMappingsTask.java
+++ b/src/main/java/net/fabricmc/loom/task/MigrateMappingsTask.java
@@ -102,7 +102,7 @@ public class MigrateMappingsTask extends AbstractLoomTask {
 		try {
 			TinyTree currentMappings = mappingsProvider.getMappings();
 			TinyTree targetMappings = getMappings(mappings);
-			migrateMappings(project, extension.getMinecraftMappedProvider(), inputDir, outputDir, currentMappings, targetMappings);
+			migrateMappings(project, extension, extension.getMinecraftMappedProvider(), inputDir, outputDir, currentMappings, targetMappings);
 			project.getLogger().lifecycle(":remapped project written to " + outputDir.toAbsolutePath());
 		} catch (IOException e) {
 			throw new IllegalArgumentException("Error while loading mappings", e);
@@ -160,8 +160,8 @@ public class MigrateMappingsTask extends AbstractLoomTask {
 		}
 	}
 
-	private static void migrateMappings(Project project, MinecraftMappedProvider minecraftMappedProvider,
-										Path inputDir, Path outputDir, TinyTree currentMappings, TinyTree targetMappings
+	private static void migrateMappings(Project project, LoomGradleExtension extension, MinecraftMappedProvider minecraftMappedProvider,
+			Path inputDir, Path outputDir, TinyTree currentMappings, TinyTree targetMappings
 	) throws IOException {
 		project.getLogger().info(":joining mappings");
 
@@ -184,6 +184,13 @@ public class MigrateMappingsTask extends AbstractLoomTask {
 
 		mercury.getClassPath().add(minecraftMappedProvider.getMappedJar().toPath());
 		mercury.getClassPath().add(minecraftMappedProvider.getIntermediaryJar().toPath());
+
+		if (extension.isForge()) {
+			mercury.getClassPath().add(minecraftMappedProvider.getSrgJar().toPath());
+			mercury.getClassPath().add(minecraftMappedProvider.getForgeMappedJar().toPath());
+			mercury.getClassPath().add(minecraftMappedProvider.getForgeIntermediaryJar().toPath());
+			mercury.getClassPath().add(minecraftMappedProvider.getForgeSrgJar().toPath());
+		}
 
 		mercury.getProcessors().add(MercuryRemapper.create(mappingSet));
 

--- a/src/main/java/net/fabricmc/loom/util/Constants.java
+++ b/src/main/java/net/fabricmc/loom/util/Constants.java
@@ -75,6 +75,7 @@ public class Constants {
 		public static final String FORGE_INSTALLER = "forgeInstaller";
 		public static final String FORGE_UNIVERSAL = "forgeUniversal";
 		public static final String FORGE_DEPENDENCIES = "forgeDependencies";
+		public static final String FORGE_NAMED = "forgeNamed";
 		public static final String MAPPING_CONSTANTS = "mappingsConstants";
 		public static final String UNPICK_CLASSPATH = "unpick";
 

--- a/src/main/java/net/fabricmc/loom/util/SourceRemapper.java
+++ b/src/main/java/net/fabricmc/loom/util/SourceRemapper.java
@@ -209,6 +209,9 @@ public class SourceRemapper {
 
 			if (extension.isForge()) {
 				m.getClassPath().add(extension.getMinecraftMappedProvider().getSrgJar().toPath());
+				m.getClassPath().add(extension.getMinecraftMappedProvider().getForgeMappedJar().toPath());
+				m.getClassPath().add(extension.getMinecraftMappedProvider().getForgeIntermediaryJar().toPath());
+				m.getClassPath().add(extension.getMinecraftMappedProvider().getForgeSrgJar().toPath());
 			}
 
 			Set<File> files = project.getConfigurations()

--- a/src/main/java/net/fabricmc/loom/util/srg/SpecialSourceExecutor.java
+++ b/src/main/java/net/fabricmc/loom/util/srg/SpecialSourceExecutor.java
@@ -35,8 +35,10 @@ import java.util.stream.Collectors;
 import java.util.zip.ZipEntry;
 
 import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.output.NullOutputStream;
 import org.gradle.api.Project;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.logging.LogLevel;
 import org.zeroturnaround.zip.ZipUtil;
 
 import net.fabricmc.loom.LoomGradleExtension;
@@ -83,8 +85,14 @@ public class SpecialSourceExecutor {
 			spec.setClasspath(specialSourceCp);
 			spec.workingDir(workingDir.toFile());
 			spec.setMain("net.md_5.specialsource.SpecialSource");
-			spec.setStandardOutput(System.out);
-			spec.setErrorOutput(System.out);
+
+			// if running with INFO or DEBUG logging
+			if (project.getGradle().getStartParameter().getLogLevel().compareTo(LogLevel.LIFECYCLE) < 0) {
+				spec.setStandardOutput(System.out);
+			} else {
+				spec.setStandardOutput(NullOutputStream.NULL_OUTPUT_STREAM);
+				spec.setErrorOutput(NullOutputStream.NULL_OUTPUT_STREAM);
+			}
 		}).rethrowFailure().assertNormalExitValue();
 
 		Files.deleteIfExists(stripped);

